### DISCUSSION TITLE
Render semi-transparent and non transparent objects seperately.

### DISCUSSION THREAD

**Discussion by BlackMoons** [FakeFishGames/Barotrauma#12985] [Ideas and suggestions]
To greatly improve draw call batching, Have you considered rendering transparent and non transparent objects separately?

Objects without transparency (Alpha test can be multi-sampled to allow nice anti-aliased edges around alpha cutouts) can be rendered with texture sort enabled in monogame to greatly improve batching, using the depth buffer to do proper depth sorting even when doing anti-aliased edges.

Then you can render semi-transparent objects in a 2nd pass, sorted by z-depth. As the majority of objects in Barotrauma are non-transparent, this should greatly improve batching. The cost would be each semi-transparent object would need to be flagged, either by being scanned when loaded or in the XML as a flag for the sprite. 

Alpha to coverage can even be used to render semi-transparent objects with order independence, though you'll give up some discrete levels of alpha blending to do so, so it might not be suitable for all graphics. Other order independent blending methods can also work depending on the desired effect, generally with some attention to how they are used.

**Comment by neirenoir** [comment]
I noticed in the ragdoll editor as well that joints in semi-transparent ragdolls will always render over each other. You could always design your creature around this and avoid much sprite overlapping, but it would be nice if there was a way to define the opacity of a group, then render that group in a separate framebuffer that then gets the applies the alpha settings when rendering the rest of the world.
